direct_database_sql_test.go

```diff
@@ -0,0 +1,104 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package direct provides a driver that connects directly to AlloyDB without
+// traversing the server side proxy while also enabling Auto IAM AuthN.
+package alloydbconn_test
+
+// [START alloydb_databasesql_connect_iam_authn_direct]
+import (
+	"context"
+	"database/sql"
+	"database/sql/driver"
+	"fmt"
+
+	"github.com/jackc/pgx/v4"
+	"github.com/jackc/pgx/v4/stdlib"
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+)
+
+// Register the driver as "alloydb-direct" to use the customer alloydbDirect.
+func init() {
+	sql.Register("alloydb-direct", &alloydbDirect{})
+}
+
+// alloydbDirect demonstrates how to implement the database/sql/driver.Driver
+// interface to enable Auto IAM AuthN without using the Go Connector. Most
+// users will want to copy this driver code and adjust its token generation to
+// suit their needs. As written the driver uses Application Default Credentials
+// which will work for ~80% of use cases.
+type alloydbDirect struct{}
+
+func authToken() (*oauth2.Token, error) {
+	ts, err := google.DefaultTokenSource(context.Background(),
+		"https://www.googleapis.com/auth/cloud-platform",
+	)
+	if err != nil {
+		return nil, err
+	}
+	return ts.Token()
+}
+
+func (p *alloydbDirect) Open(name string) (driver.Conn, error) {
+	config, err := pgx.ParseConfig(name)
+	if err != nil {
+		return nil, err
+	}
+	// Fetch the auth token and update the configuration's password before
+	// attempting to connect. This ensures all connections will use a fresh
+	// OAuth2 token.
+	tok, err := authToken()
+	if err != nil {
+		return nil, err
+	}
+	config.Password = tok.AccessToken
+
+	dbURI := stdlib.RegisterConnConfig(config)
+	if err != nil {
+		return nil, err
+	}
+	return stdlib.GetDefaultDriver().Open(dbURI)
+}
+
+// connectDirectDatabaseSQLAutoIAMAuthN establishes a connection to your
+// database using database/sql with a small wrapper driver that provides a
+// fresh OAuth2 token for every new connection. This approach enables you to
+// connect to your AlloyDB instance directly without the AlloyDB Go Connector
+// while still using Auto IAM Authentication.
```

Comment:
I wonder if we should still be calling this Automatic IAM Authentication? I guess in concept to the user it is still automatic

Reply:
Introducing another term would be confusing. So I propose we call this "Automatic IAM Authentication on the Direct Path" or something similar.

Reply:
WDYT?

Reply:
I think we may as well continue with it.